feat(global-filters): Introduce global generic filters

iker-barriocanal opened the pull request:
This PR adds support for global generic filters.

Generic filters are a new schema of inbound filters with generic conditions to avoid creating a new filter for each use case.
Project configs already support [project-specific generic filters](https://github.com/getsentry/relay/commit/ee3036cd33d6458f18c11edf4f5ba985bd03d14b). Global generic filters have been designed to be extendable, future-compatible, independent, and flexible to work well in isolation but also with project generic filters.

For event processing, Relay uses both the project config of the event and the global config. Before running the filters, Relay will merge the generic filters from the two configs and apply these to the event. If there's a match relay will drop the event and generate an outcome, or forward the event to the next processing step if there isn't.

## Why

Currently, adding a new generic filter results in an increase in the size of project configs (`config_size * number_of_filters`). This can result in a significant increase in memory usage in all the services that store project configs (`increase_per_project * number_of_projects`), especially Relay. Incorporating generic filters into global configs allows Relay to apply these filters similarly with minimal overhead.

There are additional, less important goals that this approach also achieves: not increasing network usage, not impacting project config computation time, etc.

## Previous work

Generic inbound filters were introduced in project configs in https://github.com/getsentry/relay/commit/ee3036cd33d6458f18c11edf4f5ba985bd03d14b. They are currently not enabled on `sentry` as supporting these filters in the global config was prioritized first.

## Protocol

Global generic filters have the same protocol as project generic filters, but they exist in the global config. A new optional `filters` key is added as a top-level key in global configs. This object expects the following properties:
- `version`: the version of global configs, for future-compatibility, as a `u16`.
- `filters`: a list of generic filters sorted by descending priority, with the same protocol as with the project generic filters.

Example of a full protocol:
```json
{
	// ... other global config keys ...
	"filters": {
		"version": 1,
		"filters": [
			{
				"id": "sample-errors",
				"isEnabled": true,
				"condition": {
					"op": "eq",
					"name": "event.exceptions",
					"value": "Sample Error"
				}
			}
		]
	}
}
```

## Decisions and trade-offs

### Protocol

Supporting external Relays with future compatibility to make updates safely:
- Introduced `version` to skip filters on a version not supported by Relay.
- When filters are unsupported, Relay forwards events upstream without extracting metrics. This allows the next updated Relay in the chain supporting the filters to run filtering and, in case there's a decision to keep the event, extract metrics from it.

Relay merges the configs prioritizing the project config. Each filter has an `isEnabled` flag to maximize flexibility and allow:
- Defining a filter for a single project in its config.
- Defining a filter for all projects in the global config.
- Defining a filter for most projects by enabling it in the global config and disabling it in the applicable project configs.
- Defining a filter for some projects by disabling it in the global config and enabling it in the applicable project configs.
This is the simplest alternative to support all the use cases.

### Relay implementation details

Relay types the filters internally as an ordered map with custom (de)serialization, keeping the order in the protocol, to benefit from:
- `O(1)` retrieval of filters given the ID.
- Removing duplicated filters.
Removing duplicates is a prerequisite to merge the filters from both configs properly. Since the list of filters in the payload is expected to be sorted by descending priority, only the first occurrence of a filter is selected and the rest are discarded during deserialization. Discarded filters are not forwarded downstream.

`filters` is typed as an `ErrorBoundary` instead of an `Option`:
- Errors during deserialization indicate breaking changes. Relay skips metric extraction and dynamic sampling in these cases, and the next relay in chain deals with them.
  - Note: processing relays will run dynamic sampling and extract metrics even if filters are broken.
- An empty entry in a payload is deserialized as an empty list of filters, effectively a noop.
- A list of filters in a supported version is appropriately deserialized and applied.

iker-barriocanal committed:
feat(global-filters): Introduce global generic filters

iker-barriocanal commented on relay-filter/src/config.rs:
````diff
@@ -248,13 +251,117 @@ impl GenericFilterConfig {
+/// ```
+#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct GenericFiltersConfig {
````
Since this is used in `should_filter`, which is public, should also be public.

iker-barriocanal commented on relay-filter/src/config.rs:
```diff
@@ -328,15 +482,209 @@ impl FiltersConfig {
+/// See [`CombinedFiltersConfigIter`] for details on how to iterate easily
+/// through the applicable filters.
+#[derive(Copy, Clone, Debug)]
+pub(crate) struct CombinedFiltersConfig<'a> {
```
The module structure is not great, which required the `pub(crate)` here. Refactoring to a proper structure is out of scope of this PR though, I may approach it in the future.

olksdr replied:
why not just `pub`?

iker-barriocanal replied:
It's not needed. `CombinedFiltersConfig` (and `CombinedFiltersConfigIter`) is only used internally in `relay-filter`, so exposing it seems unnecessary. I'm ok making it `pub` if you have a preference for it.

Dav1dde replied:
I'd prefer to keep this scoped to the module, this has quite some heavy logic attached to it and shouldn't be used outside of the module.

olksdr replied:
Then I would keep it private all together.

iker-barriocanal replied:
It's used in `generic.rs`, so it's either `pub(crate)` or `pub`. I'm going to leave it as `pub(crate)` unless there's a strong opinion on a different suggestion. We can revisit the module structure in the future.

Dav1dde replied:
Can you just move it to the generic module? Kind of fits there.

iker-barriocanal replied:
Actually yes, moved it in [b57c122](https://github.com/getsentry/relay/pull/3161/commits/b57c12245eaaafd914ab1719e97e42dfd65356b8).

iker-barriocanal commented on tests/integration/test_metrics.py:
```diff
@@ -1880,3 +1879,121 @@ def test_metric_bucket_encoding_dynamic_global_config_option(
+    relay.send_transaction(project_id, transaction)
+
+    tx, _ = tx_consumer.get_event()
+    assert tx is not None
```
`get_event` already asserts the transaction is `None`, but that's not super clear so I added the check here for readability.

iker-barriocanal commented on tests/integration/test_metrics.py:
```diff
@@ -1851,8 +1852,6 @@ def test_metric_bucket_encoding_dynamic_global_config_option(
             namespace: "array"
         }
 
-    print(mini_sentry.global_config)
```
Note to self: investigate why this is not caught by automation.

Dav1dde replied:
Oops, I think that was me 🤣

iker-barriocanal commented on relay-filter/src/config.rs:
````diff
@@ -248,13 +251,117 @@ impl GenericFilterConfig {
+/// Deserialization:
+///
+/// ```
+/// # use relay_filter::GenericFiltersConfig;
````
I decided to add the (de)serialization tests as doctests as the description above may not be 100% clear and I was going to add tests anyway.

iker-barriocanal committed:
Update changelog

iker-barriocanal committed:
rm development comments

olksdr commented on relay-filter/src/config.rs:
```diff
@@ -248,13 +251,117 @@ impl GenericFilterConfig {
+        deserialize_with = "deserialize_generic_filters",
+        serialize_with = "serialize_generic_filters"
+    )]
+    pub filters: IndexMap<String, GenericFilterConfig>,
```
This might be already answered: Do we definitely want to keep the order of the rules? Does the order of the rules have some special meaning?

iker-barriocanal replied:
Rules are sorted by priority, so the order of the rules is preserved, and `IndexMap` ensures that (see [docs](https://docs.rs/indexmap/latest/indexmap/map/struct.IndexMap.html#order)):

> The key-value pairs have a consistent order that is determined by the sequence of insertion and removal calls on the map.

olksdr commented on relay-filter/src/config.rs:
```diff
@@ -248,13 +251,117 @@ impl GenericFilterConfig {
+        deserialize_with = "deserialize_generic_filters",
+        serialize_with = "serialize_generic_filters"
+    )]
+    pub filters: IndexMap<String, GenericFilterConfig>,
```
Do we want to skip serialization when the `filters` is empty ?  Or your customer serialization function is handling this?

iker-barriocanal replied:
Good question!
- In `GenericFiltersConfig`, an empty filter map is serialized into an empty list.
- In the global config, filters are not serialized on `GenericFiltersConfig` deserialization errors or when it's the default value (no filters and `version` is 0 (default value of `u16`)).

Expanding on the second point: we shouldn't skip serialization on empty filters without considering the version, as a higher version can disable metric extraction in downstream relays. Sending empty filters doesn't cause deserialization errors.

Dav1dde replied:
@olksdr Do you mean skipping the entire object or just the `filters` field here? I think skipping the field we can always do.

Re skipping the object (which means we also skip the version) is a bit tricky and I'd probably default to always serializing the object if we got one from upstream.

olksdr replied:
~Yeah, entire object~ scratch that..  To skip `filters` here .

iker-barriocanal replied:
Done in [45e2f94](https://github.com/getsentry/relay/pull/3161/commits/45e2f94c76d625acc63521cd0f31e3308ffa365f) (including a new doctest).

olksdr commented on relay-filter/src/config.rs:
```diff
@@ -328,15 +482,208 @@ impl FiltersConfig {
+                    }
+                    continue;
+                }
+            }
```
Does this mean we have one filter from project config and then from global config, and next iteration the same? 
Don't we want to iterate first all project config filters and only then all global config filters? 

Or maybe I"m getting this wrong?

iker-barriocanal replied:
The `loop` iterates through all project filters first, until `project_index` is too big and thus has evaluated all project filters. Then, the iterator moves on to global filters, and repeats the same approach. I added a small comment in [950325d](https://github.com/getsentry/relay/pull/3161/commits/950325dff6dc769f46dfcc6bb4122962dacd3f42) to provide a bit more visibility.

I acknowledge this piece of code is not straightforward to understand, so I'm open to suggestions on how to make it more readable!

olksdr reviewed (commented):
Seems like all your tests have max 2 filters in the config. Can we add at least 5 to each of the location and make sure to test that they are applied in according to the rules describes in this PR?

TBS1996 commented on relay-filter/src/config.rs:
```diff
@@ -328,15 +482,208 @@ impl FiltersConfig {
+/// See [`CombinedFiltersConfigIter`] for details on how to iterate easily
+/// through the applicable filters.
+#[derive(Copy, Clone, Debug)]
+pub(crate) struct CombinedFiltersConfig<'a> {
```
so currently we have used the `Dynamic` prefix for these combined global and project types. I'm fine with changing to `Combined` but then we should change the other instances too.

iker-barriocanal replied:
Good point! Addressed in [950325d](https://github.com/getsentry/relay/pull/3161/commits/950325dff6dc769f46dfcc6bb4122962dacd3f42).

TBS1996 commented on relay-dynamic-config/src/global.rs:
```diff
@@ -23,6 +25,9 @@ pub struct GlobalConfig {
     pub quotas: Vec<Quota>,
+    /// Configuration for global inbound filters.
+    #[serde(skip_serializing_if = "skip_default_error_boundary")]
+    pub filters: ErrorBoundary<GenericFiltersConfig>,
```
it seems similar to `MeasurementsConfig` in it's form. So maybe we should also add errorboundary to the measurementsconfig in another PR if it makes sense to do it in genericfiltersconfig 🤔

iker-barriocanal replied:
What's the uncovered use case that adding `ErrorBoundary` to `MeasurementsConfig` covers?

iker-barriocanal committed:
address feedback

Changes include:
- Rename `CombinedFiltersConfig` -> `DynamicFiltersConfig` (same for the
  iterator) to align with `DynamicMeasurementsConfig`.
- Add unit test covering multiple filters from project and global
  filters.
- Add a brief comment in the iterator's loop to provide a bit more
  clarity into what the `loop` does.

Dav1dde commented on Cargo.toml:
```diff
@@ -26,6 +26,7 @@ insta = { version = "1.31.0", features = ["json", "redactions", "ron"] }
 hash32 = "0.3.1"
 hashbrown = "0.14.3"
 itertools = "0.10.5"
+indexmap = "2.0.0"
```
We could bump the dependency to `2.2.3` (latest)

iker-barriocanal replied:
Happy to do that in a follow-up if the new version introduces some nice features. I don't want to increase the scope of this PR as it's already big and complex enough.

jjbayer replied:
How would using the latest version increase the scope of this PR? It's a new dependency anyway?

iker-barriocanal replied:
2.0.0 -> 2.2.3 should have no breaking changes, but there could be e.g. uncaught panics. If this is the case and we identify this PR as the root cause, narrowing the cause further to the dependency is more complicated. Separating it in a different PR is easier to spot, and I don't mind the extra bit of work.

Dav1dde commented on relay-dynamic-config/src/global.rs:
```diff
@@ -23,6 +25,9 @@ pub struct GlobalConfig {
     /// Quotas that apply to all projects.
     #[serde(skip_serializing_if = "Vec::is_empty")]
     pub quotas: Vec<Quota>,
+    /// Configuration for global inbound filters.
```
```suggestion
    /// Configuration for global inbound filters.
    ///
    /// The configuration in the global config is later merged with the more specific
    /// configuration from the project state.
```

Something along the lines that this is only have the truth, what do you think?

Dav1dde commented on relay-dynamic-config/src/global.rs:
```diff
@@ -23,6 +25,9 @@ pub struct GlobalConfig {
     pub quotas: Vec<Quota>,
+    /// Configuration for global inbound filters.
+    #[serde(skip_serializing_if = "skip_default_error_boundary")]
+    pub filters: ErrorBoundary<GenericFiltersConfig>,
```
Do we want to be a bit more specific here, like `inbound_filters`?

iker-barriocanal replied:
We can rename the variable and set a serde `rename`, but this seemed too much to me. What do you think?

The reason to keep `filters` in the protocol is because "inbound" is a `sentry` product term and Relay doesn't have to know about it. If new non-inbound filters are introduced, Relay may run them at this point.

Dav1dde replied:
For my taste it's a bit too generic on the protocol level, but if there is a good reason to do it 👍

iker-barriocanal replied:
In project configs, the field is `filterSettings`. I'm ok with updating the protocol to `inboundFilters`, but that also generates a consistency mismatch. What do you think?

Dav1dde replied:
I'd prefer `inboundFilter` but maybe worth just bringing up with the team tomorrow.

iker-barriocanal replied:
For consistency with the rest of Relay, we're sticking with `filters`.

Dav1dde commented on relay-dynamic-config/src/global.rs:
```diff
@@ -46,6 +58,21 @@ impl GlobalConfig {
+    /// Returns the version of generic inbound filters.
+    ///
+    /// If the filters failed to deserialize, [`u16::MAX`] is returned.
+    pub fn generic_filters_version(&self) -> u16 {
```
Any reason this doesn't just return an `Option`, the special case with `u16::MAX` seems non obvious at first and I think it makes it easier to understand if the caller has that logic.

iker-barriocanal replied:
Agree, removed this method in [b57c122](https://github.com/getsentry/relay/pull/3161/commits/b57c12245eaaafd914ab1719e97e42dfd65356b8).

Dav1dde commented on relay-dynamic-config/src/global.rs:
```diff
@@ -46,6 +58,21 @@ impl GlobalConfig {
+    /// Returns the version of generic inbound filters.
+    ///
+    /// If the filters failed to deserialize, [`u16::MAX`] is returned.
+    pub fn generic_filters_version(&self) -> u16 {
```
The other method is called `filters` we should align the naming.

iker-barriocanal replied:
Removed this method in [b57c122](https://github.com/getsentry/relay/pull/3161/commits/b57c12245eaaafd914ab1719e97e42dfd65356b8).

Dav1dde commented on relay-filter/src/config.rs:
```diff
@@ -248,13 +251,117 @@ impl GenericFilterConfig {
 
+impl PartialEq for GenericFilterConfig {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id
```
While this impl does make sense it is scary, it may not be immediately obvious that the impl only compares the IDs, e.g. if you want to compare an enabled to a disabled filter with the same ID you get `true`

iker-barriocanal replied:
As comparing two configs is useful for tests, I added the full implementation just for `cfg=test` in [45e2f94](https://github.com/getsentry/relay/pull/3161/commits/45e2f94c76d625acc63521cd0f31e3308ffa365f).

Dav1dde commented on relay-filter/src/config.rs:
```diff
@@ -248,13 +251,117 @@ impl GenericFilterConfig {
-#[derive(Clone, Debug, Default, Serialize, Deserialize)]
-pub(crate) struct GenericFiltersConfig {
+///
+/// # Deserialization
```
Awesome docs!

We should doctests + examples more often.

Dav1dde commented on relay-filter/src/config.rs:
```diff
@@ -264,10 +371,57 @@ impl GenericFiltersConfig {
+        where
+            A: de::SeqAccess<'de>,
+        {
+            let mut filters = IndexMap::new();
```
`SeqAccess` gives you a `size` hint which we can pass on to `IndexMap::with_capacity`, something like `IndexMap::with_capacity(seq.size_hint().unwrap_or(0))`

Dav1dde commented on relay-filter/src/config.rs:
```diff
@@ -264,10 +371,57 @@ impl GenericFiltersConfig {
+    S: Serializer,
+{
+    let mut seq = serializer.serialize_seq(Some(index_map.len()))?;
+    for (_, filter) in index_map.iter() {
```
```suggestion
    for filter in index_map.values() {
```

Dav1dde commented on relay-filter/src/config.rs:
```diff
@@ -328,15 +482,215 @@ impl FiltersConfig {
+    pub fn new(
+        project_filters: &'a GenericFiltersConfig,
+        global_filters: Option<&'a GenericFiltersConfig>,
+        max_version: u16,
```
This confused me at first, this is the supported version, initially I thought it was the `global_version.max(project_version)`.

```suggestion
        supported_version: u16,
```

Dav1dde commented on relay-filter/src/config.rs:
```diff
@@ -328,15 +482,215 @@ impl FiltersConfig {
+    }
+}
+
+/// Iterator over [`GenericFilterConfig`] of a project and global config.
```
Good docs, but I think they should be on `DynamicFiltersConfig` to have a higher chance to be seen. I'd probably never check the docs on the return value of `into_iter()`.

iker-barriocanal replied:
Is the best place the `DynamicFiltersConfig` struct or the `into_iter` function of the `IntoIterator` implementation?

iker-barriocanal replied:
In the new approach in [a6db146](https://github.com/getsentry/relay/pull/3161/commits/a6db146fd2cbf1c653f989bbb05790e425e8023d), I added the relevant documentation in the function generating the iterator.

Dav1dde commented on relay-filter/src/config.rs:
```diff
@@ -328,15 +482,215 @@ impl FiltersConfig {
+    /// Index of the next filter in global config to evaluate.
+    global_index: usize,
+    /// Filters that have been evaluated, either yielded or ignored.
+    evaluated: HashSet<&'a String>,
```
```suggestion
    evaluated: HashSet<&'a str>,
```

jjbayer commented on relay-dynamic-config/src/global.rs:
```diff
@@ -23,6 +25,9 @@ pub struct GlobalConfig {
     pub quotas: Vec<Quota>,
+    /// Configuration for global inbound filters.
+    #[serde(skip_serializing_if = "skip_default_error_boundary")]
+    pub filters: ErrorBoundary<GenericFiltersConfig>,
```
If we treat `ErrorBoundary::Err` like `None`, why not use `default_on_error`? That could even make the getters below unnecessary.

iker-barriocanal replied:
We need a way to tell the difference between not being able to deserialize the configs (e.g. unsupported) or not getting any config at all. This influences the decision to disable metric extraction. I tried to document this as bullet points in the last trade-off in the PR description, let me know if I could make it easier to read.

Dav1dde commented on relay-filter/src/config.rs:
```diff
@@ -328,15 +482,215 @@ impl FiltersConfig {
+        if self.config.max_version < self.config.project_filters.version {
+            return false;
+        }
+        if let Some(gc) = self.config.global_filters {
```
Shorter and imo easier to understand:

```rs
self.config.project_filters.version <= self.config.max_version
    && self.config.global_filters.map_or(true, |gc| gc.version <= self.config.max_version)
```

iker-barriocanal replied:
Good suggestion! I'm doing this refactor but keeping the `is_version_supported`, as removing this indirection decreases the readability and want to keep the docstrings.

Dav1dde commented on relay-filter/src/config.rs:
```diff
@@ -328,15 +482,215 @@ impl FiltersConfig {
+        // `continue` restarts the loop, so global filters aren't evaluated
+        // until all project filters are.
+        loop {
+            if self.project_index < self.config.project_filters.filters.len() {
```
I think this check should be in `self.next_project_filter`

iker-barriocanal replied:
This check outside the function allows skipping project filters, useful when the iterator has already exhausted them. Moving the check to the function would require returning an enum to know whether the evaluated filter should be yielded, skipped or all filters are exhausted.

(same response as in the comment below)

Dav1dde commented on relay-filter/src/config.rs:
```diff
@@ -328,15 +482,215 @@ impl FiltersConfig {
+            }
+
+            if let Some(global_filters) = self.config.global_filters {
+                if self.global_index < global_filters.filters.len() {
```
If we also move this check into `self.next_global_filter`, the `if let Some()` is no longer necessary.

And we can write the entire loop as:

```rs
self.next_project_filter().or_else(self.next_global_filter)
```

iker-barriocanal replied:
This check outside the function allows skipping project filters, useful when the iterator has already exhausted them. Moving the check to the function would require returning an enum to know whether the evaluated filter should be yielded, skipped or all filters are exhausted.

(same response as in the comment above)

jjbayer commented on relay-filter/src/config.rs:
```diff
@@ -264,10 +371,57 @@ impl GenericFiltersConfig {
+                if filters.contains_key(&filter.id) {
+                    continue;
+                }
+                filters.insert(filter.id.clone(), filter);
```
nit: 
```suggestion
                if !filters.contains_key(&filter.id) {
                    filters.insert(filter.id.clone(), filter);
                }  
```

Dav1dde commented on relay-filter/src/config.rs:
```diff
@@ -328,15 +482,215 @@ impl FiltersConfig {
+
+        // Mark the filter as evaluated first. If the filter is disabled, it
+        // should not be applied from global configs.
+        let newly_evaluated = self.evaluated.insert(id);
```
`newly_evaluated == false` is impossible right? Because it is the map key.

jjbayer commented on relay-filter/src/config.rs:
```diff
@@ -248,13 +251,117 @@ impl GenericFilterConfig {
+    /// the same id. See struct docs for more details.
+    #[serde(
+        deserialize_with = "deserialize_generic_filters",
+        serialize_with = "serialize_generic_filters"
```
nit: See https://serde.rs/field-attrs.html#with

```suggestion
        with = "module_name",
```

iker-barriocanal replied:
I'm not sure if this increases readability but applied the change in [45e2f94](https://github.com/getsentry/relay/pull/3161/commits/45e2f94c76d625acc63521cd0f31e3308ffa365f).

jjbayer commented on relay-filter/src/config.rs:
```diff
@@ -264,10 +371,57 @@ impl GenericFiltersConfig {
 #[derive(Clone, Debug, Default, Serialize, Deserialize)]
 #[serde(rename_all = "camelCase")]
-pub struct FiltersConfig {
+pub struct ProjectFiltersConfig {
```
What's the reason for this rename?

iker-barriocanal replied:
This PR introduces some additional concepts around filters, and `FiltersConfig` is somewhat ambiguous about what the actual config is about. In the future, reading `FiltersConfig` would make me check the docs or go to the definition, whereas I believe `ProjectFiltersConfig` won't.

If there are opinions against the rename, I'm happy to revert it.

jjbayer commented on relay-filter/src/config.rs:
```diff
@@ -328,15 +482,215 @@ impl FiltersConfig {
+/// See [`DynamicFiltersConfigIter`] for details on how to iterate easily
+/// through the applicable filters.
+#[derive(Copy, Clone, Debug)]
+pub(crate) struct DynamicFiltersConfig<'a> {
```
Doesn't have to be exactly this name, but `Generic` should be part of it IMO.

```suggestion
pub(crate) struct DynamicGenericFiltersConfig<'a> {
```

Dav1dde commented on relay-filter/src/config.rs:
```diff
@@ -328,15 +482,215 @@ impl FiltersConfig {
+            return Some(filter);
+        }
+
+        Some(filter)
```
```suggestion
        if filter.condition.is_some() {
            return Some(filter)
        } else {
            self.config.global_filters.and_then(|filters| filters.get(id))
            // .filter(|filter| filter.condition.is_some()) <- I think that makes sense but would be new behaviour
        }
```

Dav1dde commented on relay-filter/src/config.rs:
```diff
@@ -328,15 +482,215 @@ impl FiltersConfig {
+            return None;
+        }
+
+        Some(filter)
```
```suggestion
        self.evaluated.insert(id).then_some(filter)
```

(Maybe I like combinators too much)

Dav1dde commented on relay-filter/src/config.rs:
```diff
@@ -328,15 +482,215 @@ impl FiltersConfig {
+    }
+}
+
+impl<'a> Iterator for DynamicFiltersConfigIter<'a> {
```
Iterator can be fused, we should implement it according to the [guidelines](https://develop.sentry.dev/rust/#iterators).

iker-barriocanal replied:
TIL fuse iterator, nice!

iker-barriocanal replied:
Done in [45e2f94](https://github.com/getsentry/relay/pull/3161/commits/45e2f94c76d625acc63521cd0f31e3308ffa365f).

Dav1dde commented on tests/integration/test_filters.py:
```diff
@@ -282,3 +282,37 @@ def test_ignore_transactions_filters_are_applied(
         assert event["transaction"] == transaction_name
+
+
+def test_global_filters_drop_events(
```
An additional test that has a more complex scenario of combining global and project state would be nice.

Can be one test that tests a bunch of things like:
- Disabled in the global config but enabled in project state
- Enabled in global config, disabled in project state
- Enabled in both
- Disabled in both

Could also include version mismatches and differences in the tests but you already have separate tests for that 👍

iker-barriocanal replied:
I'll cover these cases with unit tests, as they are validating the iterator's implementation logic.

Edit: these edge cases should be covered with existing and newly added unit tests in [45e2f94](https://github.com/getsentry/relay/pull/3161/commits/45e2f94c76d625acc63521cd0f31e3308ffa365f).

Dav1dde replied:
Fair enough, although that should also be nicely doable with pytest and parameterized tests.

iker-barriocanal replied:
It's possible, but the unit tests are more scoped, easier to spot, faster to run, and less surface area to be flaky 😄

Dav1dde commented on tests/integration/test_metrics.py:
```diff
@@ -1880,3 +1879,121 @@ def test_metric_bucket_encoding_dynamic_global_config_option(
+    metrics_consumer = metrics_consumer()
+    transactions_consumer = transactions_consumer()
+
+    mini_sentry.global_config.pop("filters")
```
This shouldn't be necessary, do the tests fail without?

iker-barriocanal replied:
Not sure if I understand the question. This test validates that missing global filters keep metric extraction enabled, and this line removes filters from the global config.

Dav1dde replied:
But you never added filters in the first place, I am probably missing something?

iker-barriocanal replied:
Yep, I added them to the fixture in `mini_sentry`. All tests using that fixture now get the following config:

```txt
"filters": {"version": 1, "filters": []},
```

Dav1dde commented on relay-server/src/services/processor.rs:
```diff
@@ -1274,12 +1274,22 @@ impl EnvelopeProcessorService {
         dynamic_sampling::normalize(state);
-        event::filter(state)?;
+        let global_config = self.inner.global_config.current();
+        event::filter(state, global_config.filters())?;
```
I think I prefer `event::filter` takes a reference to the entire global config and then just picks the filters out of it, gives us a little bit more type safety.

iker-barriocanal replied:
`relay-dynamic-config` depends on `relay-filter` so passing the full configs isn't an option and we're limited to just passing the versions, unfortunately.

Dav1dde replied:
Isnt `event::filter` in `relay-server`?

iker-barriocanal replied:
Oh right hehe yeah I agree, and applied in [8384897](https://github.com/getsentry/relay/pull/3161/commits/83848978c571ce5ad9a82cb47064a3bebc318a2a).

Dav1dde commented on relay-server/src/services/processor.rs:
```diff
@@ -1274,12 +1274,22 @@ impl EnvelopeProcessorService {
+        // An inbound filter applied in another up-to-date relay in chain may
+        // need to drop the event, and there should not be metrics from dropped
+        // events.
+        let supported_generic_filters = state.project_state.config.filter_settings.generic.version
```
This logic should be hidden away in the `relay-filters` module and `GENERIC_FILTERS_VERSION` shouldn't be public.

And maybe a crazy idea: what if `event::filter` returns an additional struct/enum containing that info?

iker-barriocanal replied:
Generally, replacing booleans with enums in function arguments increases readability. In this case, it's a supported/unsupported decision as the function return type, so I think a boolean is a better choice. What do you think?

Dav1dde replied:
I like the enum 👍

iker-barriocanal replied:
I've looked into it, and still think the use-case is simple enough to deal with a boolean. Let me know if you're strongly opinionated on this and feel free to block the PR in that case.

jjbayer commented on relay-filter/src/config.rs:
```diff
@@ -328,15 +482,215 @@ impl FiltersConfig {
+            }
+
+            if let Some(global_filters) = self.config.global_filters {
+                if self.global_index < global_filters.filters.len() {
```
Might just be me but I find `if ... else if ... else` much easier to understand than multiple continue statements.

```suggestion
            } else if let Some(global_filters) = self.config.global_filters && self.global_index < global_filters.filters.len() {
```

iker-barriocanal replied:
This seems to be an unstable feature, see https://github.com/rust-lang/rust/issues/53667. rust-analyzer also warns about the unnecessary `else` block.

iker-barriocanal replied:
Followed a different and simpler approach in [a6db146](https://github.com/getsentry/relay/pull/3161/commits/a6db146fd2cbf1c653f989bbb05790e425e8023d).

jjbayer commented on relay-filter/src/config.rs:
```diff
@@ -328,15 +482,215 @@ impl FiltersConfig {
+                }
+            }
+
+            return None;
```
```suggestion
            } else {
               return None;
            }
```

iker-barriocanal replied:
The suggestion results in a different behavior: the iterator skips `None` return values and goes to another loop, instead of returning `None`. On the other hand, I think that a similar block for project and global filters makes the block easier to understand.

iker-barriocanal replied:
Followed a different and simpler approach in [a6db146](https://github.com/getsentry/relay/pull/3161/commits/a6db146fd2cbf1c653f989bbb05790e425e8023d).

jjbayer commented on relay-filter/src/config.rs:
```diff
@@ -328,15 +482,215 @@ impl FiltersConfig {
+}
+
+impl<'a> IntoIterator for DynamicFiltersConfig<'a> {
+    type Item = &'a GenericFilterConfig;
```
I believe we can simplify the iterator logic by a lot if we had an `Item` type `MergedGenericFilterConfig<'a>` here, that offers methods to retrieve the merged config. Then the iterator would be something like

```rust
match project_filters.next() {
    Some(pf) => return MergedConfig::new(pf, global_filters.get(pf.id)),
    None => match global_filters.next() {
        Some(gf) => return MergedConfig::new(gf, None),
        None => None,
    }
}
```

iker-barriocanal replied:
Assuming the merging logic happens in `MergedGenericFilterConfig` or `MergedConfig`, how would the indexes for the next filter be updated? IIUC this can't happen in the `Item` of the iterator.

On the other hand, I don't like creating an abstraction that returns an abstraction over filters. As the client of the API, I just want to easily get the filters back in an easy single step.

jjbayer replied:
You wouldn't even need to track any indexes, the code above would already give you the body for the `next()` implementation.

Since there's a lot of comments on this PR, what do you think about a live discussion / some pair programming tomorrow?

iker-barriocanal replied:
Followed the simplified approach in [a6db146](https://github.com/getsentry/relay/pull/3161/commits/a6db146fd2cbf1c653f989bbb05790e425e8023d).

iker-barriocanal committed:
feedback

iker-barriocanal committed:
more feedback

iker-barriocanal committed:
fix doc links

iker-barriocanal committed:
some more feedback

iker-barriocanal committed:
fix lint :)

iker-barriocanal committed:
simplify iterator

iker-barriocanal committed:
Merge branch 'master' into iker/feat/generic-inbound-filters

iker-barriocanal commented on relay-filter/src/generic.rs:
```diff
@@ -27,44 +26,178 @@ fn matches(event: &Event, condition: Option<&RuleCondition>) -> bool {
+impl<'a> Iterator for DynamicGenericFiltersConfigIter<'a> {
+    type Item = MergedFilter<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
```
Creating two iterators before `impl Iterator` makes the iterator implementation easier to understand, but adds unneeded complexity on typing these iterators in the struct: exact type of the iterator including the closure of the filtering that must be known at compile time and creating yet another struct is not worth it.

iker-barriocanal commented on relay-filter/src/generic.rs:
```diff
@@ -27,44 +26,178 @@ fn matches(event: &Event, condition: Option<&RuleCondition>) -> bool {
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some((id, filter)) = self.project.get_index(self.project_index) {
+            self.project_index += 1;
+            let merged = merge_filters(filter, self.global.and_then(|gf| gf.get(id)));
```
The biggest simplification from the previous revision is to not check `is_enabled` or `condition`. This increases the complexity of the iterator implementation enough to make it hard to understand, and dealing with these outside the iterator is not too hard. Thus, I'm not adding these checks now, and we can consider extending the abstraction in the future to deal with these, if necessary.

Dav1dde commented on relay-filter/src/generic.rs:
```diff
@@ -27,44 +26,178 @@ fn matches(event: &Event, condition: Option<&RuleCondition>) -> bool {
+        loop {
+            let Some((id, filter)) = self
+                .global
+                .and_then(|filters| filters.get_index(self.global_index))
```
```suggestion
            let (id, filter) = self.global?.get_index(self.global_index)?;
```

Dav1dde commented on relay-dynamic-config/src/global.rs:
```diff
@@ -23,6 +24,15 @@ pub struct GlobalConfig {
+        deserialize_with = "default_on_error",
+        skip_serializing_if = "Option::is_none"
+    )]
+    pub filters: Option<GenericFiltersConfig>,
```
Wasn't there an `ErrorBoundary` before?

iker-barriocanal replied:
Yes, but it's no longer needed.

We only extract metrics in the following cases:
- In processing Relays. We don't want to limit that further.
- When events are going to be dropped due to dynamic sampling. Since there's a sampling decision outside generic filters, we don't want to limit this either.
In other cases, the events are forwarded upstream. This means that `Option` is good enough.

iker-barriocanal replied:
Even if not needed, restored it in [d7790ea](https://github.com/getsentry/relay/pull/3161/commits/d7790eab2cb3ba91d9269a1bddb685314db7b632) to make the difference in the future.

Dav1dde commented on relay-filter/src/config.rs:
```diff
@@ -248,13 +251,177 @@ impl GenericFilterConfig {
 }
 
+#[cfg(test)]
+impl PartialEq for GenericFilterConfig {
```
Should be able to just derive the `PartialEq`

Dav1dde commented on relay-filter/src/config.rs:
```diff
@@ -248,13 +251,177 @@ impl GenericFilterConfig {
+        skip_serializing_if = "IndexMap::is_empty",
+        with = "generic_filters_custom_serialization"
+    )]
+    pub filters: IndexMap<String, GenericFilterConfig>,
```
I wonder if we should make this a newtype instead and hide the `IndexMap` which should make it impossible to construct an invalid index map. 

Can also implement Serialize/Deserialize directly on the newtype (cannot forget to add the annotation)

Would also simplify tests, could just be constructed from a list/vector of filter configs again.

iker-barriocanal replied:
Done in [d7790ea](https://github.com/getsentry/relay/pull/3161/commits/d7790eab2cb3ba91d9269a1bddb685314db7b632).

Dav1dde commented on relay-filter/src/generic.rs:
```diff
@@ -27,44 +26,178 @@ fn matches(event: &Event, condition: Option<&RuleCondition>) -> bool {
+}
+
+#[derive(Debug, Default, PartialEq)]
+struct MergedFilter<'a> {
```
nit: I think `GenericFilterConfigRef` is a better name

jjbayer commented on Cargo.toml:
```diff
@@ -26,6 +26,7 @@ insta = { version = "1.31.0", features = ["json", "redactions", "ron"] }
 hash32 = "0.3.1"
 hashbrown = "0.14.3"
 itertools = "0.10.5"
+indexmap = "2.0.0"
```
nit: 
```suggestion
indexmap = "2.0.0"
itertools = "0.10.5"
```

iker-barriocanal committed:
some more feedback

Dav1dde commented on relay-filter/src/config.rs:
```diff
@@ -416,34 +409,60 @@ pub struct GenericFiltersConfig {
+    }
+
+    /// Returns whether the map is empty.
+    pub fn is_empty(&self) -> bool {
```
With the deref this shouldnt be necessary

Dav1dde replied:
Needed for serde annotation

iker-barriocanal replied:
It's used for skip serialization in serde.

Dav1dde commented on relay-filter/src/config.rs:
```diff
@@ -416,34 +409,60 @@ pub struct GenericFiltersConfig {
+    fn from(filters: Vec<GenericFilterConfig>) -> Self {
+        let mut map = IndexMap::with_capacity(filters.len());
+        for filter in filters {
+            map.insert(filter.id.clone(), filter.clone());
```
```
map.insert(filter.id.clone(), filter);
```

Dav1dde replied:
You need to check if there is not already an item in the map, same thing you do in the Deserialize, order is important.

iker-barriocanal committed:
missing checks on From

jjbayer commented on relay-dynamic-config/src/global.rs:
```diff
@@ -46,6 +55,21 @@ impl GlobalConfig {
+    /// Returns the generic inbound filters.
+    pub fn filters(&self) -> Option<&GenericFiltersConfig> {
+        match &self.filters {
+            ErrorBoundary::Err(_) => None,
```
In this case, we filter nothing, right?

Not a strong opinion on this, because we can always bump the metrics extraction version, but Isn't the following scenario still relevant:

1. We make a breaking change to the filter config.
2. Outdated external relays stop filtering, and extract metrics from everything that was previously filtered out.

I'm also on board with just deciding that we'll never make a breaking change, but in that case, why do we need the error boundary?

iker-barriocanal replied:
Should be addressed in [bfb8657](https://github.com/getsentry/relay/pull/3161/commits/bfb8657364cfc5fa9739b625bf72f78aec987d20).

Dav1dde commented on relay-filter/src/config.rs:
```diff
@@ -249,25 +253,258 @@ impl GenericFilterConfig {
     /// Returns true if the filters are not declared.
     pub fn is_empty(&self) -> bool {
-        self.filters.is_empty()
+        self.filters.0.is_empty()
```
```suggestion
        self.filters.is_empty()
```

Dav1dde commented on relay-filter/src/generic.rs:
```diff
@@ -27,32 +25,153 @@ fn matches(event: &Event, condition: Option<&RuleCondition>) -> bool {
+        loop {
+            let Some((id, filter)) = self.global?.get_index(self.global_index) else {
+                return None;
+            };
```
```suggestion
            let (id, filter) = self.global?.get_index(self.global_index)?; 
```

Dav1dde commented on relay-server/src/services/processor/event.rs:
```diff
@@ -273,6 +274,7 @@ pub fn finalize<G: EventProcessing>(
 
 pub fn filter<G: EventProcessing>(
     state: &mut ProcessEnvelopeState<G>,
+    global_config: Arc<GlobalConfig>,
```
```suggestion
    global_config: &GlobalConfig,
```

Dav1dde reviewed (approved):
An integration test which tests an invalid global config would be nice to make sure DS is skipped even with a valid project config.

iker-barriocanal committed:
disable metrics on unsupported/broken generic global filters

iker-barriocanal committed:
Merge branch 'master' into iker/feat/generic-inbound-filters

iker-barriocanal committed:
add more tests

iker-barriocanal commented on relay-server/src/services/processor.rs:
```diff
@@ -1274,23 +1274,41 @@ impl EnvelopeProcessorService {
+        // events.
+        // In processing relays, always extract metrics to avoid losing them.
+        let supported_generic_filters = self.inner.global_config.current().filters.is_ok()
+            && relay_filter::are_generic_filters_supported(
```
Unfortunately, passing the global config here creates a circular dependency with crates.

iker-barriocanal committed:
fix test

iker-barriocanal committed:
sample-rate-test 0

jjbayer commented on relay-dynamic-config/src/global.rs:
```diff
@@ -46,6 +55,21 @@ impl GlobalConfig {
+    }
+}
+
+fn skip_generic_filters(filters_config: &ErrorBoundary<GenericFiltersConfig>) -> bool {
```
nit

```suggestion
fn is_err_or_empty(filters_config: &ErrorBoundary<GenericFiltersConfig>) -> bool {
```

jjbayer commented on relay-filter/src/config.rs:
```diff
@@ -249,12 +253,164 @@ impl GenericFilterConfig {
+///
+/// # Deserialization
+///
+/// `filters` is expected to be a [`Vec<GenericfilterConfig>`].
```
Should this say something like `filters` is expected to be a sequence of `GenericfilterConfig>`  in the serialized format?

jjbayer commented on relay-filter/src/config.rs:
```diff
@@ -264,10 +420,91 @@ impl GenericFiltersConfig {
+    }
+}
+
+impl From<Vec<GenericFilterConfig>> for GenericFiltersMap {
```
```suggestion
#[cfg(test)]
impl From<Vec<GenericFilterConfig>> for GenericFiltersMap {
```

iker-barriocanal replied:
It's also used in the doctest.

jjbayer commented on relay-filter/src/generic.rs:
```diff
@@ -27,32 +33,151 @@ fn matches(event: &Event, condition: Option<&RuleCondition>) -> bool {
+fn merge_generic_filters<'a>(
+    project: &'a GenericFiltersConfig,
+    global: Option<&'a GenericFiltersConfig>,
+    #[cfg(test)] version: u16,
```
nit

```suggestion
    #[cfg(test)] max_supported_version: u16,
```

jjbayer commented on relay-filter/src/generic.rs:
```diff
@@ -4,17 +4,23 @@
 ///
 /// If the version in the project config is higher, no generic filters are applied.
-pub const VERSION: u16 = 1;
+const VERSION: u16 = 1;
```
nit

```suggestion
const MAX_SUPPORTED_VERSION: u16 = 1;
```

jjbayer commented on relay-filter/src/generic.rs:
```diff
@@ -27,32 +33,151 @@ fn matches(event: &Event, condition: Option<&RuleCondition>) -> bool {
+) -> GenericFilterConfigRef<'a> {
+    GenericFilterConfigRef {
+        id: primary.id.as_str(),
+        is_enabled: primary.is_enabled,
```
This means we cannot create "opt-out" filters that are enabled by default via the the global config. Not a blocker  but IMO the better solution would be to make `is_enabled` an `Option<bool>` in `ProjectFilterConfig` and use the global value when it it `None`.

iker-barriocanal replied:
If a filter is enabled by default in the global config, it's still possible to override it from the project config:
- Project config filters are evaluated first, along with the filters in the global configs, and have priority in this merging step.
- When going through the global config filters, those filters that exist in the project filters aren't evaluated again.
Let me know if there's a way to make this more clear.

jjbayer replied:
You're right, we don't need `is_enabled` to be `Option<bool>` because we can just omit the project-side config if we want the global default to work, and add a project-side config if we want to override. Sorry for the confusion.

jjbayer commented on relay-server/src/services/processor.rs:
```diff
@@ -1274,23 +1274,41 @@ impl EnvelopeProcessorService {
 
-        // Remember sampling decision, before it is reset in `dynamic_sampling::sample_envelope_items`.
-        let sampling_should_drop = state.sampling_result.should_drop();
+        let mut sampling_should_drop = false;
```
Why is this declared outside of the `if`?

iker-barriocanal replied:
- The value is used in the metric timer below, and only overwritten inside the `if`.
- If the `if` doesn't run, the event is kept and forwarded to the next relay, thus the default value `false`.

jjbayer commented on tests/integration/test_metrics.py:
```diff
@@ -1880,3 +1878,172 @@ def test_metric_bucket_encoding_dynamic_global_config_option(
+        },
+        # Config is valid, but filters aren't supported
+        {
+            "version": "65535",
```
Don't we expect an integer here?

```suggestion
            "version": 65535,
```

jjbayer commented on tests/integration/test_metrics.py:
```diff
@@ -1880,3 +1878,172 @@ def test_metric_bucket_encoding_dynamic_global_config_option(
+        assert tx is not None
+    else:
+        assert mini_sentry.captured_events.get() is not None
+    metrics_consumer.assert_empty()
```
I thought processing relays always extract metrics?

iker-barriocanal replied:
They do! The test was inaccurate and there's something wrong with `assert_empty`. I adapted the test in [28ed597](https://github.com/getsentry/relay/pull/3161/commits/28ed597458dcf483fb76e8baa287b00b7a5c682b) to reflect the reality better.

iker-barriocanal committed:
fix tests

iker-barriocanal committed:
fix docs

Dav1dde commented on relay-server/src/services/processor.rs:
```diff
@@ -1274,23 +1274,41 @@ impl EnvelopeProcessorService {
-        if self.inner.config.processing_enabled() || sampling_should_drop {
-            self.extract_metrics(state)?;
-        }
+        if self.inner.config.processing_enabled() || supported_generic_filters {
```
Shouldn't this be:

```suggestion
        if self.inner.config.processing_enabled() && supported_generic_filters {
```

iker-barriocanal replied:
Nope: processing relays should always run dynamic sampling, even if generic filters are unsupported.

Dav1dde replied:
Can Managed Relays also have processing enabled?

Dav1dde replied:
Nvm that wouldn't make sense.

jjbayer replied:
They shouldn't, but we've seen customers do it. IMO `Processing` should be a mode, just like `Managed`.

Dav1dde commented on relay-server/src/services/processor.rs:
```diff
@@ -1274,23 +1274,41 @@ impl EnvelopeProcessorService {
+        // need to drop the event, and there should not be metrics from dropped
+        // events.
+        // In processing relays, always extract metrics to avoid losing them.
+        let supported_generic_filters = self.inner.global_config.current().filters.is_ok()
```
I really think `event::filter` should return this, this is filter related logic and just hard to read here.

or hide it behind some function in the `event` module, then you can also pass the global config.

iker-barriocanal replied:
Makes sense. I'll address this in a follow-up so that this PR can already land.

Dav1dde commented on relay-server/src/services/processor.rs:
```diff
@@ -1274,23 +1274,41 @@ impl EnvelopeProcessorService {
 
-        // Remember sampling decision, before it is reset in `dynamic_sampling::sample_envelope_items`.
-        let sampling_should_drop = state.sampling_result.should_drop();
+        let mut sampling_should_drop = false;
```
This doesn't need to be here and also not be mutable.

Just always use `sampling_should_drop = state.sampling_result.should_drop();`

iker-barriocanal replied:
`dynamic_sampling::sample_envelope_items` resets the value, see comment below. This change is not coming from this PR.

Dav1dde replied:
Ah nvm sorry I misread the indentation

Dav1dde reviewed (approved):
As noted before I think we can move the `supported_generic_filters` to a better place. Other than that :shipit:
